server/src/main/java/keywhiz/service/providers/ClientAuthenticator.java

```diff
@@ -0,0 +1,172 @@
+
+    Optional<Client> possibleClientFromName = possibleClientName.flatMap((name) -> {
+      if (clientAuthConfig.typeConfig().useCommonName()) {
+        return clientDAOReadOnly.getClientByName(name);
+      } else {
+        return Optional.empty();
+      }
+    });
+
+    Optional<Client> possibleClientFromSpiffeId = possibleClientSpiffeId.flatMap((name) -> {
+      if (clientAuthConfig.typeConfig().useSpiffeId()) {
+        return clientDAOReadOnly.getClientBySpiffeId(name);
+      } else {
+        return Optional.empty();
+      }
+    });
+
+    // If the name and SPIFFE id both defined a client, make sure that they defined the same
+    // client! (Note that if retrieving clients by common name or SPIFFE ID is disabled, this
+    // check will not reject a certificate with a mismatched CN and SPIFFE URI.)
+    if (possibleClientFromName.isPresent() && possibleClientFromSpiffeId.isPresent()) {
+      if (!possibleClientFromName.get().equals(possibleClientFromSpiffeId.get())) {
+        throw new NotAuthorizedException(String.format(
+            "Input principal's CN and SPIFFE ID correspond to different clients! (cn = %s, spiffe = %s)",
+            possibleClientFromName.get().getName(), possibleClientFromSpiffeId.get().getName()));
+      }
+    } else if (possibleClientFromName.isEmpty() && possibleClientFromSpiffeId.isEmpty()) {
+      // Create missing clients if configured to do so (client name must be present)
+      return handleMissingClient(createMissingClient, possibleClientName.orElse(""),
+          possibleClientSpiffeId.orElse(""));
+    }
+
+    // Retrieve whichever of the clients is set (if both are present, the earlier check guarantees
+    // that they contain the same client).
+    Client client = possibleClientFromName
+        .or(() -> possibleClientFromSpiffeId)
+        .orElseThrow(() -> new IllegalStateException(
+            "Unable to identify client, and fallback code in server did not handle this case"));
+
+    // Record that this client has been retrieved
+    clientDAOReadWrite.sawClient(client, principal);
+    if (client.isEnabled()) {
+      return Optional.of(client);
+    } else {
+      logger.warn("Client {} authenticated but disabled via DB", client);
+      return Optional.empty();
+    }
+  }
+
+  private Optional<Client> handleMissingClient(boolean createMissingClient, String name,
+      String spiffeId) {
+    if (createMissingClient && !name.isEmpty()) {
+      /*
+       * If a client is seen for the first time, authenticated by certificate, and has no DB entry,
+       * then a DB entry is created here. The client can be disabled in the future by flipping the
+       * 'enabled' field.
+       */
+      long clientId = clientDAOReadWrite.createClient(name, "automatic",
+          "Client created automatically from valid certificate authentication", spiffeId);
+      return clientDAOReadWrite.getClientById(clientId);
```

Comment:
There's a small behavior change here: the original authenticator returned
`return Optional.of(clientDAOReadWrite.getClientById(clientId).get());`
which would throw a `NoSuchElementException` if the client couldn't be created. However, the only situations I can think of where the client creation fails would be if the client already existed (not the case here, barring race conditions) or when the database is down; in either case, cleanly denying authentication rather than throwing an exception is probably better.

Reply:
makes sense to me